File: perl-programming/examples/arrays/reverse_polish_calculator.pl
#!/usr/bin/perl
use strict;
use warnings;

my @stack;
while (1) {

    print '$ ';
    my $in = <STDIN>;
    chomp $in;

    if ($in eq "q") { last; }

    if ($in eq "c") { 
            pop @stack; 
            next;
            }   # fetch the last value
    if ($in eq "*") { 
            my $x = pop(@stack);
            my $y = pop(@stack);
            push(@stack, $x*$y);
            next;
            }
    if ($in eq "+") { 
            my $x = pop(@stack);
            my $y = pop(@stack);
            push(@stack, $x + $y);
            next;
            }
    if ($in eq "/") {
            my $x = pop(@stack);
            my $y = pop(@stack);
            push(@stack, $y /  $x); 
            next;
            }
    if ($in eq "-") { 
            my $x = pop(@stack);
            my $y = pop(@stack);
            push(@stack, $y - $x);
            next;
            }
    if ($in eq "=") { 
            print pop(@stack), "\n"; 
            next;
            }
    push @stack, $in;
}

File: books/python-programming/src/examples/lists/reverse_polish_calculator.py
stack = []

print("x = eXit, s = Show, [+-*/=]")
while True:
    val = input(':')

    if val == 's':
        print(stack)
        continue

    if val == 'x':
        break

    if val == '+':
        a = stack.pop()
        b = stack.pop()
        stack.append(a+b)
        continue

    if val == '-':
        a = stack.pop()
        b = stack.pop()
        stack.append(a-b)
        continue

    if val == '*':
        a = stack.pop()
        b = stack.pop()
        stack.append(a*b)
        continue

    if val == '/':
        a = stack.pop()
        b = stack.pop()
        stack.append(a/b)
        continue

    if val == '=':
        print(stack.pop())
        continue

    stack.append(float(val))
